File: ssh_session_store.py
"""SSH session data persistence for SOSterm."""

import configparser
import glob
import json
import os
import re
import uuid
from dataclasses import dataclass, field, asdict
from typing import List, Optional


@dataclass
class SSHGroup:
    id: str = ""
    name: str = ""
    color: str = ""
    expanded: bool = True

    def __post_init__(self):
        if not self.id:
            self.id = f"g-{uuid.uuid4().hex[:12]}"


@dataclass
class SSHSession:
    id: str = ""
    group_id: str = ""
    name: str = ""
    host: str = ""
    port: int = 22
    username: str = ""
    auth_method: str = "key"  # "key" or "password"
    identity_file: str = ""
    startup_command: str = ""
    color: str = ""

    def __post_init__(self):
        if not self.id:
            self.id = f"s-{uuid.uuid4().hex[:12]}"

    def display_name(self) -> str:
        if self.name:
            return self.name
        label = ""
        if self.username:
            label += f"{self.username}@"
        label += self.host
        if self.port and self.port != 22:
            label += f":{self.port}"
        return label

    def build_command(self, default_identity_file="") -> list:
        cmd = ["ssh"]
        if self.port and self.port != 22:
            cmd += ["-p", str(self.port)]
        # Use session-specific key, or fall back to the global default
        identity = self.identity_file if self.identity_file else default_identity_file
        if identity:
            cmd += ["-i", os.path.expanduser(identity)]
        target = ""
        if self.username:
            target = f"{self.username}@{self.host}"
        else:
            target = self.host
        cmd.append(target)
        if self.startup_command:
            cmd.append(self.startup_command)
        return cmd


_CONFIG_DIR = os.path.expanduser("~/.config/SOSterm")
_SESSIONS_FILE = os.path.join(_CONFIG_DIR, "ssh_sessions.json")


class SSHSessionStore:
    """Load/save SSH sessions and groups from JSON."""

    def __init__(self, path=None):
        self._path = path or _SESSIONS_FILE
        self._groups: List[SSHGroup] = []
        self._sessions: List[SSHSession] = []
        self.load()

    # --- Persistence ---

    def load(self):
        if not os.path.exists(self._path):
            self._groups = []
            self._sessions = []
            return
        try:
            with open(self._path, "r") as f:
                data = json.load(f)
            self._groups = [SSHGroup(**g) for g in data.get("groups", [])]
            self._sessions = [SSHSession(**s) for s in data.get("sessions", [])]
        except (json.JSONDecodeError, TypeError, KeyError):
            self._groups = []
            self._sessions = []

    def save(self):
        os.makedirs(os.path.dirname(self._path), exist_ok=True)
        data = {
            "version": 1,
            "groups": [asdict(g) for g in self._groups],
            "sessions": [asdict(s) for s in self._sessions],
        }
        with open(self._path, "w") as f:
            json.dump(data, f, indent=2)

    # --- Group CRUD ---

    def groups(self) -> List[SSHGroup]:
        return list(self._groups)

    def get_group(self, group_id: str) -> Optional[SSHGroup]:
        for g in self._groups:
            if g.id == group_id:
                return g
        return None

    def add_group(self, group: SSHGroup):
        self._groups.append(group)
        self.save()

    def update_group(self, group: SSHGroup):
        for i, g in enumerate(self._groups):
            if g.id == group.id:
                self._groups[i] = group
                self.save()
                return
        self.add_group(group)

    def delete_group(self, group_id: str):
        self._groups = [g for g in self._groups if g.id != group_id]
        # Move sessions in this group to ungrouped
        for s in self._sessions:
            if s.group_id == group_id:
                s.group_id = ""
        self.save()

    # --- Session CRUD ---

    def sessions(self) -> List[SSHSession]:
        return list(self._sessions)

    def get_session(self, session_id: str) -> Optional[SSHSession]:
        for s in self._sessions:
            if s.id == session_id:
                return s
        return None

    def sessions_in_group(self, group_id: str) -> List[SSHSession]:
        return [s for s in self._sessions if s.group_id == group_id]

    def ungrouped_sessions(self) -> List[SSHSession]:
        return [s for s in self._sessions if not s.group_id]

    def add_session(self, session: SSHSession):
        self._sessions.append(session)
        self.save()

    def update_session(self, session: SSHSession):
        for i, s in enumerate(self._sessions):
            if s.id == session.id:
                self._sessions[i] = session
                self.save()
                return
        self.add_session(session)

    def delete_session(self, session_id: str):
        self._sessions = [s for s in self._sessions if s.id != session_id]
        self.save()

    # --- SSH config import ---

    def import_ssh_config(self, config_path=None) -> List[SSHSession]:
        """Parse ~/.ssh/config and return candidate sessions (skips wildcards)."""
        if config_path is None:
            config_path = os.path.expanduser("~/.ssh/config")
        if not os.path.exists(config_path):
            return []

        candidates = []
        existing_hosts = {(s.host, s.port, s.username) for s in self._sessions}

        try:
            with open(config_path, "r") as f:
                lines = f.readlines()
        except OSError:
            return []

        current_host = None
        current_data = {}

        for line in lines:
            line = line.strip()
            if not line or line.startswith("#"):
                continue

            # Match key-value pairs
            match = re.match(r"(\w+)\s+(.+)", line)
            if not match:
                continue

            key = match.group(1).lower()
            value = match.group(2).strip()

            if key == "host":
                # Save previous host
                if current_host and current_data:
                    self._add_candidate(current_host, current_data,
                                        existing_hosts, candidates)
                # Start new host block
                current_host = value
                current_data = {}
            elif key == "hostname":
                current_data["hostname"] = value
            elif key == "port":
                try:
                    current_data["port"] = int(value)
                except ValueError:
                    pass
            elif key == "user":
                current_data["user"] = value
            elif key == "identityfile":
                current_data["identity_file"] = value

        # Handle last block
        if current_host and current_data:
            self._add_candidate(current_host, current_data,
                                existing_hosts, candidates)

        return candidates

    def _add_candidate(self, host_alias, data, existing_hosts, candidates):
        """Add a candidate session from ssh config parsing."""
        # Skip wildcard entries
        if "*" in host_alias or "?" in host_alias:
            return

        hostname = data.get("hostname", host_alias)
        port = data.get("port", 22)
        user = data.get("user", "")

        # Dedupe against existing
        if (hostname, port, user) in existing_hosts:
            return

        session = SSHSession(
            name=host_alias if host_alias != hostname else "",
            host=hostname,
            port=port,
            username=user,
            auth_method="key" if data.get("identity_file") else "password",
            identity_file=data.get("identity_file", ""),
        )
        candidates.append(session)

    # --- Remmina import ---

    def import_remmina(self, remmina_dir=None) -> List[SSHSession]:
        """Parse Remmina .remmina files and return SSH candidate sessions."""
        if remmina_dir is None:
            # Check common locations
            for candidate_dir in [
                os.path.expanduser("~/.local/share/remmina"),
                os.path.expanduser("~/.remmina"),
            ]:
                if os.path.isdir(candidate_dir):
                    remmina_dir = candidate_dir
                    break
        if not remmina_dir or not os.path.isdir(remmina_dir):
            return []

        existing_hosts = {(s.host, s.port, s.username) for s in self._sessions}
        candidates = []

        for filepath in sorted(glob.glob(os.path.join(remmina_dir, "*.remmina"))):
            try:
                cfg = configparser.ConfigParser()
                cfg.read(filepath)
                if not cfg.has_section("remmina"):
                    continue
                r = cfg["remmina"]

                # Only import SSH connections
                if r.get("protocol", "").upper() != "SSH":
                    continue

                name = r.get("name", "")
                server = r.get("server", "")
                username = r.get("username", "")
                group_name = r.get("group", "")
                identity_file = r.get("ssh_privatekey", "")
                exec_cmd = r.get("exec", "")

                if not server:
                    continue

                # Parse host:port
                host = server
                port = 22
                if ":" in server:
                    parts = server.rsplit(":", 1)
                    host = parts[0]
                    try:
                        port = int(parts[1])
                    except ValueError:
                        pass

                # Dedupe
                if (host, port, username) in existing_hosts:
                    continue

                auth_method = "key" if identity_file else "password"

                session = SSHSession(
                    name=name,
                    host=host,
                    port=port,
                    username=username,
                    auth_method=auth_method,
                    identity_file=identity_file,
                    startup_command=exec_cmd,
                )
                # Stash group name for caller to use
                session._remmina_group = group_name
                candidates.append(session)

            except (configparser.Error, OSError):
                continue

        return candidates

File: src/sshsessionstore.cpp
#include "sshsessionstore.h"
#include "settings.h"

#include <QDir>
#include <QFile>
#include <QJsonArray>
#include <QJsonDocument>
#include <QJsonObject>
#include <QSettings>
#include <QUuid>

// --- SSHSession ---

QString SSHSession::displayName() const
{
    if (!name.isEmpty())
        return name;
    QString label;
    if (!username.isEmpty())
        label = username + "@";
    label += host;
    if (port != 0 && port != 22)
        label += ":" + QString::number(port);
    return label;
}

QStringList SSHSession::buildSSHArgs(const QString &defaultKey) const
{
    QStringList args;
    if (port != 0 && port != 22)
        args << "-p" << QString::number(port);

    QString identity = identityFile.isEmpty() ? defaultKey : identityFile;
    if (!identity.isEmpty()) {
        QString expanded = identity;
        if (expanded.startsWith("~/"))
            expanded = QDir::homePath() + expanded.mid(1);
        args << "-i" << expanded;
    }

    QString target = username.isEmpty() ? host : (username + "@" + host);
    args << target;

    if (!startupCommand.isEmpty())
        args << startupCommand;

    return args;
}

// --- SSHSessionStore ---

SSHSessionStore::SSHSessionStore(const QString &path, QObject *parent)
    : QObject(parent)
    , m_path(path.isEmpty() ? Settings::configDir() + "/ssh_sessions.json" : path)
{
    load();
}

QString SSHSessionStore::generateId(const QString &prefix)
{
    return prefix + "-" + QUuid::createUuid().toString(QUuid::Id128).left(12);
}

void SSHSessionStore::load()
{
    m_groups.clear();
    m_sessions.clear();

    QFile file(m_path);
    if (!file.open(QIODevice::ReadOnly))
        return;

    QJsonDocument doc = QJsonDocument::fromJson(file.readAll());
    if (!doc.isObject())
        return;

    QJsonObject root = doc.object();

    for (const auto &val : root["groups"].toArray()) {
        QJsonObject obj = val.toObject();
        SSHGroup g;
        g.id = obj["id"].toString();
        g.name = obj["name"].toString();
        g.color = obj["color"].toString();
        g.expanded = obj.value("expanded").toBool(true);
        m_groups.append(g);
    }

    for (const auto &val : root["sessions"].toArray()) {
        QJsonObject obj = val.toObject();
        SSHSession s;
        s.id = obj["id"].toString();
        s.groupId = obj["group_id"].toString();
        s.name = obj["name"].toString();
        s.host = obj["host"].toString();
        s.port = obj.value("port").toInt(22);
        s.username = obj["username"].toString();
        s.authMethod = obj["auth_method"].toString();
        s.identityFile = obj["identity_file"].toString();
        s.startupCommand = obj["startup_command"].toString();
        s.color = obj["color"].toString();
        m_sessions.append(s);
    }
}

void SSHSessionStore::save()
{
    QDir().mkpath(QFileInfo(m_path).absolutePath());

    QJsonArray groupsArr;
    for (const auto &g : m_groups) {
        QJsonObject obj;
        obj["id"] = g.id;
        obj["name"] = g.name;
        obj["color"] = g.color;
        obj["expanded"] = g.expanded;
        groupsArr.append(obj);
    }

    QJsonArray sessionsArr;
    for (const auto &s : m_sessions) {
        QJsonObject obj;
        obj["id"] = s.id;
        obj["group_id"] = s.groupId;
        obj["name"] = s.name;
        obj["host"] = s.host;
        obj["port"] = s.port;
        obj["username"] = s.username;
        obj["auth_method"] = s.authMethod;
        obj["identity_file"] = s.identityFile;
        obj["startup_command"] = s.startupCommand;
        obj["color"] = s.color;
        sessionsArr.append(obj);
    }

    QJsonObject root;
    root["version"] = 1;
    root["groups"] = groupsArr;
    root["sessions"] = sessionsArr;

    QFile file(m_path);
    if (file.open(QIODevice::WriteOnly))
        file.write(QJsonDocument(root).toJson(QJsonDocument::Indented));
}

// --- Group CRUD ---

QVector<SSHGroup> SSHSessionStore::groups() const { return m_groups; }

SSHGroup SSHSessionStore::getGroup(const QString &groupId) const
{
    for (const auto &g : m_groups)
        if (g.id == groupId) return g;
    return {};
}

void SSHSessionStore::addGroup(const SSHGroup &group)
{
    SSHGroup g = group;
    if (g.id.isEmpty())
        g.id = generateId("g");
    m_groups.append(g);
    save();
}

void SSHSessionStore::updateGroup(const SSHGroup &group)
{
    for (auto &g : m_groups) {
        if (g.id == group.id) {
            g = group;
            save();
            return;
        }
    }
}

void SSHSessionStore::deleteGroup(const QString &groupId)
{
    // Move sessions in this group to ungrouped
    for (auto &s : m_sessions) {
        if (s.groupId == groupId)
            s.groupId.clear();
    }
    m_groups.erase(std::remove_if(m_groups.begin(), m_groups.end(),
        [&](const SSHGroup &g) { return g.id == groupId; }), m_groups.end());
    save();
}

// --- Session CRUD ---

QVector<SSHSession> SSHSessionStore::sessions() const { return m_sessions; }

SSHSession SSHSessionStore::getSession(const QString &sessionId) const
{
    for (const auto &s : m_sessions)
        if (s.id == sessionId) return s;
    return {};
}

QVector<SSHSession> SSHSessionStore::sessionsInGroup(const QString &groupId) const
{
    QVector<SSHSession> result;
    for (const auto &s : m_sessions)
        if (s.groupId == groupId) result.append(s);
    return result;
}

QVector<SSHSession> SSHSessionStore::ungroupedSessions() const
{
    QVector<SSHSession> result;
    for (const auto &s : m_sessions)
        if (s.groupId.isEmpty()) result.append(s);
    return result;
}

void SSHSessionStore::addSession(const SSHSession &session)
{
    SSHSession s = session;
    if (s.id.isEmpty())
        s.id = generateId("s");
    m_sessions.append(s);
    save();
}

void SSHSessionStore::updateSession(const SSHSession &session)
{
    for (auto &s : m_sessions) {
        if (s.id == session.id) {
            s = session;
            save();
            return;
        }
    }
}

void SSHSessionStore::deleteSession(const QString &sessionId)
{
    m_sessions.erase(std::remove_if(m_sessions.begin(), m_sessions.end(),
        [&](const SSHSession &s) { return s.id == sessionId; }), m_sessions.end());
    save();
}

// --- Import SSH Config ---

QVector<SSHSession> SSHSessionStore::importSSHConfig(const QString &configPath)
{
    QString path = configPath.isEmpty()
        ? QDir::homePath() + "/.ssh/config" : configPath;

    QFile file(path);
    if (!file.open(QIODevice::ReadOnly | QIODevice::Text))
        return {};

    QVector<SSHSession> candidates;
    SSHSession current;
    bool inHost = false;

    while (!file.atEnd()) {
        QString line = QString::fromUtf8(file.readLine()).trimmed();
        if (line.isEmpty() || line.startsWith('#'))
            continue;

        int sep = line.indexOf(QRegExp("\\s+"));
        if (sep < 0) continue;
        QString key = line.left(sep).toLower();
        QString value = line.mid(sep).trimmed();

        if (key == "host") {
            if (inHost && !current.host.isEmpty())
                candidates.append(current);
            current = SSHSession();
            inHost = true;
            // Skip wildcard hosts
            if (value.contains('*') || value.contains('?')) {
                inHost = false;
                continue;
            }
            current.name = value;
        } else if (inHost) {
            if (key == "hostname")
                current.host = value;
            else if (key == "port")
                current.port = value.toInt();
            else if (key == "user")
                current.username = value;
            else if (key == "identityfile") {
                current.identityFile = value;
                current.authMethod = "key";
            }
        }
    }
    if (inHost && !current.host.isEmpty())
        candidates.append(current);

    // If host wasn't explicitly set, use the Host alias
    for (auto &s : candidates) {
        if (s.host.isEmpty())
            s.host = s.name;
    }

    // Deduplicate against existing sessions
    QVector<SSHSession> result;
    for (auto &c : candidates) {
        bool dup = false;
        for (const auto &existing : m_sessions) {
            if (existing.host == c.host && existing.port == c.port
                && existing.username == c.username) {
                dup = true;
                break;
            }
        }
        if (!dup)
            result.append(c);
    }
    return result;
}

// --- Import Remmina ---

QVector<SSHSession> SSHSessionStore::importRemmina(const QString &remminaDir)
{
    QStringList searchDirs;
    if (remminaDir.isEmpty()) {
        searchDirs << QDir::homePath() + "/.local/share/remmina"
                   << QDir::homePath() + "/.remmina";
    } else {
        searchDirs << remminaDir;
    }

    QVector<SSHSession> candidates;

    for (const auto &dirPath : searchDirs) {
        QDir dir(dirPath);
        if (!dir.exists())
            continue;

        for (const auto &entry : dir.entryInfoList({"*.remmina"}, QDir::Files)) {
            QSettings ini(entry.absoluteFilePath(), QSettings::IniFormat);
            ini.beginGroup("remmina");

            QString protocol = ini.value("protocol").toString().toUpper();
            if (protocol != "SSH")
                continue;

            SSHSession s;
            s.name = ini.value("name").toString();

            QString server = ini.value("server").toString();
            // Parse host:port
            int colonIdx = server.lastIndexOf(':');
            if (colonIdx > 0) {
                bool ok;
                int port = server.mid(colonIdx + 1).toInt(&ok);
                if (ok) {
                    s.host = server.left(colonIdx);
                    s.port = port;
                } else {
                    s.host = server;
                }
            } else {
                s.host = server;
            }

            s.username = ini.value("username").toString();
            QString keyFile = ini.value("ssh_privatekey").toString();
            if (!keyFile.isEmpty()) {
                s.identityFile = keyFile;
                s.authMethod = "key";
            }
            s.startupCommand = ini.value("exec").toString();

            ini.endGroup();

            if (!s.host.isEmpty())
                candidates.append(s);
        }
    }

    // Deduplicate
    QVector<SSHSession> result;
    for (auto &c : candidates) {
        bool dup = false;
        for (const auto &existing : m_sessions) {
            if (existing.host == c.host && existing.port == c.port
                && existing.username == c.username) {
                dup = true;
                break;
            }
        }
        if (!dup)
            result.append(c);
    }
    return result;
}
